fastfield_codecs/src/compact_space/build_compact_space.rs

```diff
@@ -0,0 +1,242 @@
+    // (conversely to all data) and we can assign null to it.
+    if compact_space.is_empty() {
+        compact_space.add_blanks(
+            blank_collector
+                .drain()
+                .map(|blank| blank.blank_range())
+                .take(1),
+        );
+    }
+
+    compact_space.finish()
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+struct CompactSpaceBuilder {
+    blanks: Vec<RangeInclusive<u128>>,
+}
+
+impl CompactSpaceBuilder {
+    /// Creates a new compact space builder which will initially cover the whole space.
+    fn new() -> Self {
+        Self { blanks: vec![] }
+    }
+
+    /// Assumes that repeated add_blank calls don't overlap and are not adjacent,
+    /// e.g. [3..=5, 5..=10] is not allowed
+    ///
+    /// Both of those assumptions are true when blanks are produced from sorted values.
+    fn add_blanks(&mut self, blank: impl Iterator<Item = RangeInclusive<u128>>) {
+        self.blanks.extend(blank);
+    }
+
+    fn is_empty(&self) -> bool {
+        self.blanks.is_empty()
+    }
+
+    /// Convert blanks to covered space and assign null value
+    fn finish(mut self) -> CompactSpace {
+        // sort by start. ranges are not allowed to overlap
+        self.blanks.sort_by_key(|blank| *blank.start());
+
+        // Between the blanks
+        let mut covered_space = self
+            .blanks
+            .iter()
+            .tuple_windows()
+            .map(|(left, right)| {
+                assert!(
+                    left.end() < right.start(),
+                    "overlapping or adjacent ranges detected"
+                );
+                *left.end() + 1..=*right.start() - 1
+            })
+            .collect::<Vec<_>>();
+
+        // Outside the blanks
+        if let Some(first_blank_start) = self.blanks.first().map(RangeInclusive::start) {
+            if *first_blank_start != 0 {
+                covered_space.insert(0, 0..=first_blank_start - 1);
+            }
+        }
+
+        if let Some(last_blank_end) = self.blanks.last().map(RangeInclusive::end) {
+            if *last_blank_end != u128::MAX {
+                covered_space.push(last_blank_end + 1..=u128::MAX);
+            }
+        }
+
+        // Extend the first range and assign the null value to it.
+        let null_value = if let Some(first_covered_space) = covered_space.first_mut() {
+            // in case the first covered space ends at u128::MAX, assign null to the beginning
+            if *first_covered_space.end() == u128::MAX {
+                *first_covered_space = first_covered_space.start() - 1..=*first_covered_space.end();
+                *first_covered_space.start()
+            } else {
+                *first_covered_space = *first_covered_space.start()..=first_covered_space.end() + 1;
+                *first_covered_space.end()
+            }
+        } else {
+            covered_space.push(0..=0); // empty data case
```

Comment:
I really don't understand the edge cases... I'm not sure if this is needlessy complicated or if it is intrinsically difficult.